### DISCUSSION TITLE
write performance decrease

### DISCUSSION THREAD

**Discussion by reisman234** [rook/rook#12638] [Q&A]
I have a running healthy  rook-ceph cluster (see status below). But when I write a large file (>10Gi)  to a persistent volume, which uses a ceph-blockpool storage class, the write performance does drop to 20MB/s and lower. 

The Cluster consists of 3 nodes, each with a hardware-raid block device (36Ti), and a 1GbE network interface.

I know that for the network a 10GbE connection is recommended, but I did not expect such low transfer drops.

--- 
```
[root@rook-ceph-tools-6c78448d68-w75p9 /]# ceph status
  cluster:
    id:     bb269df9-065c-4afd-bb1c-bcea338c2653
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum a,b,c (age 8w)
    mgr: b(active, since 2h), standbys: a
    mds: 1/1 daemons up, 1 hot standby
    osd: 3 osds: 3 up (since 8w), 3 in (since 8w)
    rgw: 1 daemon active (1 hosts, 1 zones)
 
  data:
    volumes: 1/1 healthy
    pools:   12 pools, 169 pgs
    objects: 34.80k objects, 119 GiB
    usage:   358 GiB used, 109 TiB / 109 TiB avail
    pgs:     169 active+clean
 
  io:
    client:   852 B/s rd, 808 KiB/s wr, 1 op/s rd, 2 op/s wr
 ```

**Comment by reisman234** [comment]
I forgot to mention that I did not change the default configuration for the cephBlockPool, so the pool has the default replicated size of 3. Could this also be a cause of this problem?

**Comment by travisn** [comment]
With replica 3 and exactly 3 OSDs, all the OSDs will have a replica of the data, and it must be committed to all three replicas before the IO completes, so that slowdown is common, although it does sound more extreme than expected. Ceph's strength is typically where there are multiple clients. Do you see higher throughput with more clients?

**Reply by reisman234** [reply]
Thank you, @travisn, for the response.
>  Do you see higher throughput with more clients?

Do you mean with clients, more nodes which are attached to the CephClusterSpec or something else? 

With the monitoring, I found out that the three osd's are maxing out the available network interface.
![image](https://github.com/rook/rook/assets/44891585/394d244c-00bf-4a03-a4e7-9b9fee46dd96).

This behaviour is also strange because the available host NIC is a bonding of two 1xGbE. Does the CNI of the cluster require a special configuration to be aware of a bonding NICs?

For simplicity and because the stored data is not much of value, I would just disable the replication. Would there be a problem if I just change that configuration in the BlockPoolSpec on a running and used CephCluster?

**Reply by travisn** [reply]
> Do you mean with clients, more nodes which are attached to the CephClusterSpec or something else?

I'm referring to multiple client applications that are writing at the same time. The OSDs can handle more load in parallel this way.

> For simplicity and because the stored data is not much of value, I would just disable the replication. Would there be a problem if I just change that configuration in the BlockPoolSpec on a running and used CephCluster?

You mean change the `replicated.size` to `1` in the pool? That can be done at runtime and ceph will update. This could be fine for testing, but of course dangerous for production without replication.

**Comment by reisman234** [comment]
I forgot to mention, what I do for testing the setup. 

I create a PersistentVolumeClaim with the ceph-block storageClass  and a deployment with an sshd application which uses that pvc.
From a Node outside the cluster, I do now upload a large file > 10Gi to that sshd server into the storage of the mounted pvc.

To upload a file, iIuse scp tool.  This shows the upload rate, which is at first above 90Mb/s and after some time it drops and the calculated remaining time shows just 'stalled'.

**Reply by zhucan** [reply]
why not use fio tools to test rbd volume, and than show the test results?

**Reply by reisman234** [reply]
Thank you for the hint with `fio`, I didn't know that tool. 

Here is a test result. I used the test option `readwrite` because in my opinion it is similar to the scp transmission.
```
/data # fio --rw=readwrite --name=test --size=15G 
test: (g=0): rw=rw, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=psync, iodepth=1
fio-3.33
Starting 1 process
Jobs: 1 (f=1): [M(1)][100.0%][r=101MiB/s,w=101MiB/s][r=25.9k,w=25.9k IOPS][eta 00m:00s] 
test: (groupid=0, jobs=1): err= 0: pid=1597: Mon Aug 21 07:52:18 2023
  read: IOPS=10.9k, BW=42.7MiB/s (44.7MB/s)(7684MiB/180104msec)
    clat (nsec): min=560, max=13141M, avg=70003.30, stdev=13089785.55
     lat (nsec): min=610, max=13141M, avg=70113.05, stdev=13089785.56
    clat percentiles (nsec):
     |  1.00th=[     620],  5.00th=[    1096], 10.00th=[    1208],
     | 20.00th=[    1288], 30.00th=[    1336], 40.00th=[    1384],
     | 50.00th=[    1448], 60.00th=[    1528], 70.00th=[    1608],
     | 80.00th=[    1768], 90.00th=[    1960], 95.00th=[    2160],
     | 99.00th=[  544768], 99.50th=[ 1908736], 99.90th=[ 8585216],
     | 99.95th=[10158080], 99.99th=[17432576]
   bw (  KiB/s): min=    8, max=426512, per=100.00%, avg=77541.30, stdev=50198.76, samples=203
   iops        : min=    2, max=106628, avg=19385.27, stdev=12549.70, samples=203
  write: IOPS=10.9k, BW=42.6MiB/s (44.7MB/s)(7676MiB/180104msec); 0 zone resets
    clat (nsec): min=1050, max=18615M, avg=18949.62, stdev=15582785.04
     lat (nsec): min=1100, max=18615M, avg=19073.61, stdev=15582785.36
    clat percentiles (nsec):
     |  1.00th=[   1528],  5.00th=[   2160], 10.00th=[   2288],
     | 20.00th=[   2544], 30.00th=[   2576], 40.00th=[   2640],
     | 50.00th=[   2704], 60.00th=[   2768], 70.00th=[   2864],
     | 80.00th=[   3056], 90.00th=[   3920], 95.00th=[   4384],
     | 99.00th=[   6240], 99.50th=[   8256], 99.90th=[  18048],
     | 99.95th=[  22912], 99.99th=[1613824]
   bw (  KiB/s): min=  400, max=426592, per=100.00%, avg=78227.04, stdev=49812.32, samples=201
   iops        : min=   98, max=106648, avg=19556.69, stdev=12453.07, samples=201
  lat (nsec)   : 750=1.12%, 1000=0.17%
  lat (usec)   : 2=45.89%, 4=47.45%, 10=4.35%, 20=0.20%, 50=0.03%
  lat (usec)   : 100=0.03%, 250=0.10%, 500=0.12%, 750=0.04%, 1000=0.02%
  lat (msec)   : 2=0.29%, 4=0.07%, 10=0.07%, 20=0.03%, 50=0.01%
  lat (msec)   : 100=0.01%, 250=0.01%, 500=0.01%, 750=0.01%, 1000=0.01%
  lat (msec)   : 2000=0.01%, >=2000=0.01%
  cpu          : usr=2.78%, sys=6.06%, ctx=56644, majf=0, minf=166
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwts: total=1967169,1964991,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
   READ: bw=42.7MiB/s (44.7MB/s), 42.7MiB/s-42.7MiB/s (44.7MB/s-44.7MB/s), io=7684MiB (8058MB), run=180104-180104msec
  WRITE: bw=42.6MiB/s (44.7MB/s), 42.6MiB/s-42.6MiB/s (44.7MB/s-44.7MB/s), io=7676MiB (8049MB), run=180104-180104msec

Disk stats (read/write):
  rbd6: ios=29187/1908, merge=1844/5646, ticks=207672/20863272, in_queue=21070944, util=98.78%
```

**Reply by zhucan** [reply]
you need to special the bs, and the modify the numjobs and iodpth to test. maybe you need spend some time to take a look at the fio doc.

**Reply by zhucan** [reply]
https://github.com/louwrentius/fio-plot a tools help to test.

**Comment by BlaineEXE** [comment]
It's possible to calculate the maximum theoretical throughput for any link speed. 

For 1GbE, that's 1,000,000 bits per second (for all 3 replicas). Divide by 3 to get bits per second per replica. Divide by 8 to get Bytes per second. Divide by 1024 to get MB from Gb. That's only 40 MB/s theoretical throughput. And the k8s overlay network, system processes, TCP overhead, etc is going to eat up a fair percentage of that.

Still, I might expect that you could get up to 70% of theoretical -- about 25 MByte/s, but what you are experiencing isn't out of the ballpark.

**Reply by reisman234** [reply]
Thank you for your response and clarification, I didn't think the problem through completely when I started this discussion. However, I am still somewhat clueless as to why the bonding of the NICs improves the transfer somewhat.

The cluster nodes would also have Infiniband available to form their own storage network, but this is completely new territory for me.
